server/src/AccessControl.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE OverloadedStrings #-}
+-- | This module defines the kinds of permissions used in icepeak and provides
+-- functions checking for sufficient permissions for certain operations.
+module AccessControl
+       ( AccessMode (..)
+       , AuthPath (..)
+       , IcepeakClaim (..)
+       , Path
+       , allowEverything
+       , accessModeToText
+       , textToAccessMode
+       , isAuthorizedByClaim
+       ) where
+
+import           Data.Aeson ((.:), (.=))
+import qualified Data.Aeson as Aeson
+import qualified Data.List  as List
+import           Data.Text  (Text)
+
+import           Store      (Path)
+
+-- * Claim datatypes
+
+-- | Defines the structure of a JWT claim for Icepeak.
+data IcepeakClaim = IcepeakClaim
+  { icepeakClaimWhitelist :: [AuthPath]
+    -- ^ The whitelist containing all authorizations.
+  } deriving (Read, Show, Eq, Ord)
+
+data AuthPath = AuthPath
+  { authPathPrefix :: Path
+    -- ^ The prefix of all the paths to which this authorization applies.
+  , authPathModes  :: [AccessMode]
+    -- ^ The modes that are authorized on this path prefix.
+  } deriving (Read, Show, Eq, Ord)
+
+-- | Different modes for accessing the JSON store
+data AccessMode = ModeRead | ModeWrite
+  deriving (Read, Show, Eq, Ord, Enum, Bounded)
+
+
+-- | A claim that allows all operations.
+allowEverything :: IcepeakClaim
+allowEverything = IcepeakClaim [AuthPath [] [minBound..maxBound]]
+
+-- * Authorization
+
+-- | Check whether accessing the given path with the given mode is authorized by
+-- the supplied claim.
+isAuthorizedByClaim :: IcepeakClaim -> Path -> AccessMode -> Bool
+isAuthorizedByClaim claim path mode = any allows (icepeakClaimWhitelist claim) where
```

Comment:
I'd say to move `where` to the next line to be consistent with the rest of the project.